### PR TITLE
Documentation updates for v4

### LINKED ISSUE #706: Lifetime scope in Getting Started
## Description
The step-by-step guide at https://json-api-dotnet.github.io/JsonApiDotNetCore/getting-started/step-by-step.html instructs to add the following code to Startup.ConfigureServices:

```csharp
public IServiceProvider ConfigureServices(IServiceCollection services)
{
    // add the db context like you normally would
    services.AddDbContext<AppDbContext>(options =>
    { // use whatever provider you want, this is just an example
        options.UseNpgsql(GetDbConnectionString());
    }, ServiceLifetime.Transient);

    // add jsonapi dotnet core
    services.AddJsonApi<AppDbContext>();
    // ...
}
```

I wonder why `ServiceLifetime.Transient` needs to be passed. because it is not in the official EF Core 3 documentation. Official sample at https://github.com/dotnet/AspNetCore.Docs/blob/master/aspnetcore/data/ef-rp/intro/samples/cu30/Startup.cs

Unless the scope setting is required for some reason, I think it would be better to remove it, so the context can be reused within a single request. 

See also: https://stackoverflow.com/questions/37507691/entity-framework-core-service-default-lifetime.

### PR DESCRIPTION
**Pull request by bart-degreed**
Fixes #706.

### CODE REVIEW

File: docs/usage/extensibility/repositories.md [hunk 1]
```diff
@@ -1,68 +1,66 @@
...
-    services.AddScoped<IEntityRepository<Article>, AuthorizedArticleRepository>();
+    services.AddScoped<IResourceRepository<Article>, AuthorizedArticleRepository>();
     // ...
 }
 ```
 
 A sample implementation that performs data authorization might look like this.
 
-All of the methods in the DefaultEntityRepository will use the Get() method to get the DbSet<TEntity> so this is a good method to apply scoped filters such as user or tenant authorization.
+All of the methods in the DefaultResourceRepository will use the Get() method to get the DbSet<TEntity>, so this is a good method to apply scoped filters such as user or tenant authorization.
```
Comment (fdlane) [inline docs/usage/extensibility/repositories.md:17]: did you keep the `TEntity` here because on the `DbSet`?  btw, huge thank you for all this work.

Comment (bart-degreed) [reply docs/usage/extensibility/repositories.md:17]: Yes. This is just a first pass over the documentation, type parameters may change before stable release. Thoughts at https://github.com/json-api-dotnet/JsonApiDotNetCore/issues/734

File: docs/usage/resource-graph.md [hunk 2]
```diff
@@ -47,6 +46,18 @@ public void ConfigureServices(IServiceCollection services)
 }
 ```
 
+Be aware that the previous command does not inject resource definitions and service layer overrides. You can it combine with auto-discovery to register them.
```
Comment (fdlane) [inline docs/usage/resource-graph.md:49]: consider changing `You can it combine ...` to `You can combine it ...`

Comment (bart-degreed) [reply docs/usage/resource-graph.md]: nice catch!

File: docs/usage/resources/hooks.md [hunk 3]
```diff
@@ -455,9 +455,9 @@ public override void BeforeImplicitUpdateRelationship(IAffectedRelationships<Per
 }
 ```
 
-## Using Resource Hooks without EF Core
+## Using Resource Hooks without Entity Framework Core
 
-If you want to use Resource Hooks without EF Core, there are several things that you need to consider that need to be met. For any resource that you want to use hooks for:
+If you want to use Resource Hooks without Entity Framework Core, there are several things that you need to consider that need to be met. For any resource that you want to use hooks for:
 1. The corresponding resource repository must fully implement `IEntityReadRepository<TEntity, TId>`
```
Comment (fdlane) [inline docs/usage/resources/hooks.md:461]: Should this be `ResourceReadRepository` now?

File: docs/usage/resources/resource-definitions.md [hunk 4]
```diff
@@ -48,38 +47,24 @@ Content-Type: application/vnd.api+json
 ### Single Attribute
 
 ```c#
-public class ModelResource : ResourceDefinition<Model>
+public class UserResource : ResourceDefinition<User>
 {
-    protected override List<AttrAttribute> OutputAttrs()
-        => Remove(m => m.AccountNumber);
+    public UserResource(IResourceGraph resourceGraph) : base(resourceGraph)
```
Comment (fdlane) [inline docs/usage/resources/resource-definitions.md:52]: This type of change brings so much clarity to the docs for someone like me who does not spend every day in C#. Thank you

File: docs/usage/resources/hooks.md (hunk 3 above)
Comment (bart-degreed) [reply docs/usage/resources/hooks.md:461]: Yes, thanks.

File: docs/usage/resources/resource-definitions.md (hunk 4 above)
Comment (bart-degreed) [reply docs/usage/resources/resource-definitions.md]: Do you mean that the constructor is included in the docs, or that the implementation was changed?

File: docs/usage/sorting.md [hunk 5]
```diff
@@ -1,21 +1,25 @@
...
 
 ```http
 GET /api/articles?sort=-author HTTP/1.1
-Accept: application/vnd.api+json
+```
+
+## Multiple
+
+```http
+GET /api/articles?sort=author,-pageCount HTTP/1.1
```
Comment (fdlane) [inline docs/usage/sorting.md:22]: is this `sort=author&sort=-pageCount`

File: docs/usage/resources/resource-definitions.md (hunk 4 above)
Comment (fdlane) [reply docs/usage/resources/resource-definitions.md]: Just meant your use of real-world models like `User` and not using expression body syntax for examples.

File: docs/usage/sorting.md (hunk 5 above)
Comment (fdlane) [reply docs/usage/sorting.md:22]: I don't think multiple sorts in this version is working. Maybe doing something wrong but I've lost this functionality.

File: docs/usage/resources/resource-definitions.md (hunk 4 above)
Comment (bart-degreed) [reply docs/usage/resources/resource-definitions.md]: Thanks, that is great feedback. I've updated other examples to not use expression-bodied syntax too.

File: docs/usage/sorting.md (hunk 5 above)
Comment (bart-degreed) [reply docs/usage/sorting.md:22]: Thanks for reporting. Tracked at #745.

Comment (bart-degreed) [reply docs/usage/sorting.md:22]: `sort=author&sort=-pageCount` is identical to `sort=author,-pageCount`.
ASP.NET Core merges multiple parameters with the same name automatically.

Comment (fdlane) [reply docs/usage/sorting.md:22]: Thank you, I had not realized.  

Neither seems to work when trying to execute multiple sorts on the same entity.  I understand that the nested sorts are not working, I just wanted to make sure that we called out multiple sorts on same entity is not working either.

Comment (bart-degreed) [reply docs/usage/sorting.md:22]: What do you mean with nested sorts? Like I  said, #745 tracks multiple sorts on the same entity.

File: docs/usage/resources/resource-definitions.md [hunk 6]
```diff
@@ -90,13 +75,16 @@ _since v3.0.0_
...
 {
-    protected override PropertySortOrder GetDefaultSortOrder()
-        => new PropertySortOrder {
-            (t => t.Prop, SortDirection.Ascending),
-            (t => t.Prop2, SortDirection.Descending),
+    public override PropertySortOrder GetDefaultSortOrder()
+    {
+        return new PropertySortOrder
+        {
+            (account => account.LastLoginTime, SortDirection.Descending),
```
Comment (fdlane) [inline docs/usage/resources/resource-definitions.md:84]: I think this should be...

```C#
        (account => new { account.LastLoginTime }, SortDirection.Descending),
        (account => new { account.UserName }, SortDirection.Ascending)
```

Comment (bart-degreed) [reply docs/usage/resources/resource-definitions.md:84]: Are you sure? The next is in the unit tests:
```c#
        public override PropertySortOrder GetDefaultSortOrder()
            => new PropertySortOrder {
                (t => t.Prop, SortDirection.Ascending)
            };
```

Comment (fdlane) [reply docs/usage/resources/resource-definitions.md:84]: Seems to work fine for Strings, but for other data types the following error is returned.

From the detail: `The type must be a NewExpression.Example: article => new { article.Author };`

```json
{
    "errors": [
        {
            "id": "ba97f100-e76c-44b3-b20b-b1c4b2777a2c",
            "status": "500",
            "title": "An unhandled error occurred while processing this request.",
            "detail": "The expression returned by 't => Convert(t.InvoiceDate, Object)' for 'JsonApiDotNetCore.Internal.ResourceGraph' is of type System.Linq.Expressions.UnaryExpression and cannot be used to select resource attributes. The type must be a NewExpression.Example: article => new { article.Author };"
        }
    ]
}
````

Changing to the following seems to resolve the issue.

```c#
      public override PropertySortOrder GetDefaultSortOrder() => new PropertySortOrder
      {
        (t => new { t.InvoiceDate }, SortDirection.Descending)
      };
```

Comment (fdlane) [reply docs/usage/resources/resource-definitions.md:84]: Updated the test with the property `SomeDate` and added a sort and the test now fails.

```C#
    public class Model : Identifiable
    {
        [Attr] public string AlwaysExcluded { get; set; }
        [Attr] public string Password { get; set; }
        [Attr] public string Prop { get; set; }
        [Attr] public DateTime SomeDate { get; set; }
    }

    public override PropertySortOrder GetDefaultSortOrder()
            => new PropertySortOrder {
                (t => t.Prop, SortDirection.Ascending),
                (t => t.Password, SortDirection.Descending),
                (t => t.SomeDate, SortDirection.Descending)
            };
```

Do you want me to update the test with a pr?

Comment (bart-degreed) [reply docs/usage/resources/resource-definitions.md:84]: The error message you're getting is misleading. See https://github.com/json-api-dotnet/JsonApiDotNetCore/pull/754 for the fix. Would be great if you could review that. Perhaps some testing with your API for additional types may reveal move problems.

Comment (fdlane) [reply docs/usage/resources/resource-definitions.md:84]: @bart-degreed, thank you.  We are testing.

Comment (fdlane) [reply docs/usage/resources/resource-definitions.md:84]: @bart-degreed closing the loop here also and noting that #754 fixed our api sorting.  Thank you.

**Comment by bart-degreed** [comment]
@fdlane When you're done, can you please use the Review button to approve/request changes/etc. Also, for next time 'Start a review' / 'Finish review' works better than single comments, because each new comment sends out a new notification email.

**Comment by fdlane** [comment]
@bart-degreed  Certainly.  Sorry I flooded email.

**Pull request merged**